content/polymorphic-eta-expansion.md

```diff
@@ -0,0 +1,495 @@
+
+~~~ scala 
+extension (x: Int)
+  def poly[T](x: T): T = x
+// signature: (Int)[T](T): T
+
+poly(4) // polymorphic method: takes a  [T]
+poly    // monomorphic method: takes an (Int)
+~~~
+
+Note: Since typechecking is recursive, eta-expansion of a monomorphic method like `poly` can still trigger polymorphic eta-expansion, for example:
+
+~~~ scala
+val voly: Int => [T] => T => T = poly
+// poly expands to: (x: Int) => [T] => (y: T) => poly(x)[T](y)
+~~~
+
+As this feature only provides a shortcut to express already definable objects, the only impacted area is the type system.
+
+When typing a polymorphic method `m` there are three cases to consider:
+
+#### With expected type
+If the expected type is a polymorphic function taking `[T_1 <: U_1 >: L_1, ..., T_n <: U_n >: L_n]` as type parameters, `(A_1, ..., A_k)` as term parameters and returning `R`, we proceed as follows:
+
+Note: Polymorphic functions always take term parameters (but `k` can equal zero if the clause is explicit: `[T] => () => T`).
+
+1. Copies of `T_i`s are created, and replaced in `U_i`s, `L_i`s, `A_i`s and `R`, noted respectively `T'_i`, `U'_i`, `L'_i`, `A'_i` and `R'`.
+
+2. Is the expected type a polymorphic context function ?
+* 1. If yes then `m` is replaced by the following: 
+~~~ scala
+[T'_1 <: U'_1 >: L'_1, ... , T'_n <: U'_n >: L'_n] 
+  => (a_1: A'_1 ..., a_k: A'_k)
+  ?=> m[T'_1, ..., T'_n]
+~~~
+* 2. If no then `m` is replaced by the following: 
+~~~ scala
+[T'_1 <: U'_1 >: L'_1, ... , T'_n <: U'_n >: L'_n] 
+  => (a_1: A'_1 ..., a_k: A'_k)
+  => m[T'_1, ..., T'_n](a_1, ..., a_k)
+~~~
+
+3. the application of `m` is type-checked with expected type `R'`
+* 1. If it succeeds, the above is the created tree.
+* 2. If it fails, go to [Default](#Default).
+
+At 3.2. if the cause of the error is such that [Default](#Default) will never succeed, we might return that error directly, this is at the discretion of the implementation, to make errors as clear as possible.
+
+Note: Type checking will be in charge of overloading resolution, as well as term inference, so the following will work:
+~~~ scala
+def f[A](using Int)(x: A)(using String): A
+def f[B](x: B, y: B): B
+
+given i: Int = ???
+given s: String = ???
+val v: [T] => T => T = f
+// f expands to: [T'] => (t: T') => f[T'](t)
+// and then to:  [T'] => (t: T') => f[T'](using i)(t)(using s)
+
+def g[C](using C): C
+val vg: [T] => T ?=> T = g
+// g expands to: [T'] => (t: T') ?=> g[T']
+// and then to:  [T'] => (t: T') ?=> g[T'](using t)
+~~~
+
+Note: Type checking at 3. will have to recursively typecheck `m[T'_1, ..., T'_n](a_1, ..., a_k)` with expected type `R`, this can lead to further eta-expansion:
+~~~ scala
+extension [A](x: A)
+  def foo[B](y: B) = (x, y)
+
+val voo: [T] => T => [U] => U => (T, U) = foo
+// foo expands to: 
+// [T'] => (t: T') => ( foo[T'](t) with expected type [U] => U => (T', U) )
+// [T'] => (t: T') => [U'] => (u: U') => foo[T'](t)[U'](u)
+~~~
+
+#### Without expected type
```

Comment:
To be safe I would strongly suggest that there should be no eta expansion without an expected type. The reason is that missing type parameters are always inferred and we do not want to make an exception now. Type arguments are in that sense analogous to implicits.

Here's an example:
```scala
def f(x: Int)(using C)
```
Here, `f` expands to
```scala
(x: Int) => f(x)(summon[C])
```
It does _not_ expand to
```scala 
(x: Int) => (c: C) ?=> m(x)(using c)
```
Analogously,  when faced with
```scala
class C { type T }
def g(x: C)[U >: x.T]
```
`g` should expand to 
```scala
(x: C) => g(x)[x.T]
```
and not to
```scala
(x: C) => [U] => g(x)[U]
```
Granted, the second version in the polymorphic eta expansion might seem more useful, (and maybe the second version for the implicit eta expansion is deemed more useful, too). But that's not how things work. I believe the analogy between polymorphic and implicit eta expansion is too strong to throw overboard.

Generally, I have a bit mixed feelings about this but am overall positive. The downside is that it is an "exotic" feature with not a lot of use cases, and the worked out definition is very complicated. This will have a significant impact on spec complexity. Do we really need it? On the positive side, the details are worked out very well. If we can restrict eta expansion to kick in only with a polymorphic expected type, I think it's fine. Otherwise the feature interactions with general inference principles of Scala are serious enough to hold back. 

Also note that if we would decide at some point that we need eta expansion without an expected type because we have lots of convincing use cases, we can still add it later.

Reply:
I ammended the proposal as requested, I will leave this conversation as unresolved so the reasons are apparent